add driver for SDS011 active laser dust sensor

MichelRottleuthner opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
Adds a driver and SAUL integration for the SDS011 dust sensor for measuring particulate matter concentration.
<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure
Below example is only true for `pba-d-01-kw2x`, for other boards adapt as needed.
For SAUL use the default test application:
```
cd tests/saul
USEMODULE=sds011 CFLAGS="-DSDS011_PARAM_UART_DEV=UART_DEV\(1\) -DSDS011_PARAM_PWR_PIN=GPIO_PIN\(3,5\)" make BOARD=pba-d-01-kw2x all flash term
```

With the provided test:
```
cd tests/driver_sds011
CFLAGS="-DSDS011_PARAM_UART_DEV=UART_DEV\(1\) -DSDS011_PARAM_PWR_PIN=GPIO_PIN\(3,5\)" make BOARD=pba-d-01-kw2x all flash term
```
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

smlng commented on drivers/include/sds011.h:
```diff
@@ -0,0 +1,315 @@
+  * @brief   Work/sleep mode of the SDS011 sensor
+  */
+  typedef enum sds011_working_mode
+  {
```
move brace to end if previous line

smlng commented on drivers/include/sds011.h:
```diff
@@ -0,0 +1,315 @@
+ * @brief   Report mode of the SDS011 sensor
+ */
+ typedef enum sds011_reporting_mode
+ {
```
move brace to end if previous line

smlng commented on drivers/include/sds011.h:
```diff
@@ -0,0 +1,315 @@
+int sds011_init(sds011_t *dev, const sds011_params_t *params);
+
+/**
+ * @brief   Enable power suppply of SDS011 laser dust sensor
```
/ppp/pp/

smlng commented on drivers/include/sds011.h:
```diff
@@ -0,0 +1,315 @@
+void sds011_power_on(const sds011_t *dev);
+
+/**
+ * @brief   Disable power suppply of SDS011 laser dust sensor
```
dito /ppp/pp/

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+    /* if no active reporting callback is registered, UART must be enabled first */
+    if((dev->cb == NULL) &&
+       (uart_init(dev->params.uart, SDS011_UART_BAUDRATE, _rx_cb, dev) != 0)) {
+        return SDS011_ERROR;
```
missing unlock?

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+    dev->pos = 0;
+    dev->checksum = 0;
+
+    mutex_lock(&dev->cb_lock);
```
lock again?

smlng replied:
may you wanted to unlock here?

smlng replied:
oops, now I see 2 different locks, sorry about the noise here, but still unlock before error return would be good?

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+    uart_write(dev->params.uart, cmd, SDS011_FRAME_SEND_LEN);
+
+    /* wait for the isr callback to unlock the mutex */
+    mutex_lock(&dev->cb_lock);
```
and again?

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+
+/** @brief shorthand for getting a single byte property with _send_recv_cmd  */
+static int _get_property(sds011_t *dev, uint8_t *cmd, size_t len,
+                                uint8_t *p, uint8_t p_idx){
```
align indention with parentheses

smlng replied:
and missing space `) {`

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+    if (uart_init(dev->params.uart, SDS011_UART_BAUDRATE,
+                  cb == NULL ? NULL : _rx_cb,
+                  cb == NULL ? NULL : dev) != 0){
+        return SDS011_ERROR;
```
missing unlock in case of error?

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+    }
+}
+
+int sds011_get_reporting_mode(sds011_t *dev, sds011_reporting_mode_t *mode){
```
brace on new line

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+    return _get_property(dev, cmd, sizeof(cmd), mode, SDS011_DB3_IDX);
+}
+
+int sds011_set_reporting_mode(sds011_t *dev, sds011_reporting_mode_t mode){
```
brace on new line

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+    return res;
+}
+
+int sds011_read(sds011_t *dev, sds011_data_t *data){
```
dito

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,339 @@
+    return res;
+}
+
+int sds011_set_dev_id(sds011_t *dev, uint16_t sens_dev_id){
```
dito

MichelRottleuthner commented on tests/driver_sds011/main.c:
```diff
@@ -0,0 +1,215 @@
+        return 0;
+    }
+
+    if (sds011_set_reporting_mode(&dev, SDS011_RMODE_QUERY) == SDS011_OK) {
```
copy-pasta -> removed

MichelRottleuthner commented on tests/driver_sds011/main.c:
```diff
@@ -0,0 +1,215 @@
+    if (sds011_get_working_mode(&dev, &wmode) == SDS011_OK) {
+        printf("current working mode: %s\n",
+               (wmode == SDS011_WMODE_WORK) ? "WORK" : "SLEEP");
+        if (rmode != SDS011_RMODE_QUERY) {
```
fixed

leandrolanzieri commented:
@smlng ping?

smlng commented on tests/driver_sds011/main.c:
```diff
@@ -0,0 +1,245 @@
+ */
+static sds011_t dev;
+
+static char* _rmode_str(sds011_reporting_mode_t rmode) {
```
this file still does not adhere to RIOT coding style, e.g. brace should be on next line

MichelRottleuthner replied:
damn! fixed it - may i squash?

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,351 @@
+        return SDS011_ERROR;
+    }
+
+    return res;
```
will never be reached, suggest to remove the `else` part above and simply return `SDS011_ERROR` here

MichelRottleuthner replied:
done

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,351 @@
+        }
+    }
+
+    return res;
```
this is a bit different but could be changed as suggested above, also missing space after `else `

MichelRottleuthner replied:
done

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,351 @@
+        }
+    }
+
+    return res;
```
same again

MichelRottleuthner replied:
done

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,351 @@
+{
+    uint8_t cmd[SDS011_FRAME_SEND_LEN] = {0};
+    int checksum = 0;
+    int res = 0;
```
should initialised with proper `enum` value, e.g. `SDS011_ERROR` as default

MichelRottleuthner replied:
done

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,351 @@
+
+    dev->cb = NULL;
+
+    sds011_power_on(dev);
```
why? does this make sense without a callback?

MichelRottleuthner replied:
Yes because the device needs to be started up before you can request data and the driver guidelines say a device should be ready to interact with after the init call. Also, the driver (device) provides two different modes of operation. One where you need to request a measurement and one where the sensor reports periodically. The driver defaults to requesting mode so the user is not forced to register a callback to talk to the device. You can still chose to use a callback if preferred - the driver can handle both.

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,351 @@
+            return SDS011_OK;
+        }
+        else{
+            return SDS011_ERROR;
```
see above, I suggest to remove this `else` and simply return `SDS011_ERROR` by default when not successful

MichelRottleuthner replied:
done

smlng commented on drivers/sds011/sds011_saul.c:
```diff
@@ -0,0 +1,46 @@
+       res->scale = -7;
+       return 2;
+    }
+    return 0;
```
should return `ECANCELED` on error according to API doc

MichelRottleuthner replied:
done

MichelRottleuthner commented:
@smlng ping

smlng commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,345 @@
+    return res;
+}
+
+/** @brief shorthand for getting a single byte property with _send_recv_cmd  */
```
see above,

MichelRottleuthner replied:
done

smlng commented on drivers/include/sds011.h:
```diff
@@ -0,0 +1,313 @@
+    SDS011_RMODE_QUERY  = 1, /**< sensor needs to be queried */
+} sds011_reporting_mode_t;
+
+ /**
```
remove indention by 1 space

MichelRottleuthner replied:
done

smlng commented on drivers/include/sds011.h:
```diff
@@ -0,0 +1,313 @@
+ * @brief   Device descriptor definition for SDS011 Laser Dust Sensor
+ */
+typedef struct {
+    sds011_params_t   params;
```
undocumented members?

MichelRottleuthner replied:
done

MichelRottleuthner commented on drivers/include/sds011.h:
```diff
@@ -0,0 +1,313 @@
+int sds011_read(sds011_t *dev, sds011_data_t *data);
+
+/**
+ * @brief   Register callback that is executed when new measurements were
```
done

MichelRottleuthner commented on drivers/sds011/include/sds011_params.h:
```diff
@@ -0,0 +1,78 @@
+#endif
+
+/**
+ * @brief   Set default configuration parameters for the SDS011 driver
```
done

MichelRottleuthner commented on drivers/sds011/sds011.c:
```diff
@@ -0,0 +1,345 @@
+    dev->pos++;
+}
+
+/** @brief send command and wait for the first message that is sent as reply */
```
done

MichelRottleuthner commented:
@smlng everything addressed now? and may I squash?

smlng commented on drivers/sds011/include/sds011_params.h:
```diff
@@ -0,0 +1,78 @@
+#ifndef SDS011_SAUL_INFO
+#define SDS011_SAUL_INFO             { .name = "SDS011" }
+#endif
+/**@}*/
```
sorry, but this should be `/** @} */` (with spaces)

MichelRottleuthner commented:
@smlng I fixed the remaining problems reported by murdock and removed the separate thread in the test to make it fit on the small arduinos. So, finally all spaces where they should be?^^
